### PR TITLE
Text Adventure Game - Team 5 - Assignment3

### PR DESCRIPTION
**Pull request by gstandley21**

### CODE REVIEW

File: src/main/java/Action.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import javax.swing.*;
+import java.awt.event.*;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class Action {
```
Comment (njits030) [inline src/main/java/Action.java:6]: I think this class is a bit redundant. The attributes of this classed are never used inside of this class and the only thing that is done when an instance of this class is created is parsing the users command and executing it. I think it was not necessary to make a separate class to do this.

File: src/main/java/ChatClient.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        try{
+            Socket socket = new Socket(serverAddress, 59001);
+
+            out = new PrintWriter(socket.getOutputStream(), true);
+            in = new Scanner(socket.getInputStream());
+
+            while (in.hasNextLine()) {
+                String text = in.nextLine();
+
+                if (text.startsWith("TEXT")) {
```
Comment (njits030) [inline src/main/java/ChatClient.java:54]: This should have been an if/else statement and not just three if statements. The same string can't start with TEXT, ENTERNAME and NAMEOKAY. So there is no reason to be checking all three for one string.

File: src/main/java/Main.java [hunk 3]
```diff
@@ -1,5 +1,259 @@
...
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.commons.io.FileUtils;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import static java.lang.System.*;
+
 public class Main {
```
Comment (njits030) [inline src/main/java/Main.java:11]: Maybe you could've put some of these operations in other classes or you could've created an extra class. There's a lot of different things happening here right now.

File: src/main/java/Use.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public void execute() {
+        if (command.length == 2) {
+            String itemName = command[1];
+
+            Item item = Main.itemMap.get(itemName);
+            // Hack to escape if item is not held because returning breaks other things
+            boolean skipExecute = false;
+
+            if (item == null) {
```
Comment (njits030) [inline src/main/java/Use.java:20]: Should have used if and else if here.

File: docs/assignment3.md [hunk 5]
```diff
@@ -1 +1,205 @@
...
+* Updated class diagram to properly represent JSON objects 
+* Updated object diagram to represent the completed implementation and any changes that were made from the assignment 2 code
+* Added association names to the class diagram 
+* Added the chat room feature (Added **ChatClient** class) so that you can communicate with other players (use command "chat")
+* Implemented time limit feature to make the game slightly more complex
+* Added obstacles to the game using the **Obstacle** class to make the game more complex
+* Added the "use" command (implemented as the **Use** class) to the game to apply **Item** to **Obstacles** to clear them
+* Implemented persistence to allow a user to save their progress. The objective behind this feature is to allow users to begin a new game, play, exit the terminal, and then restart in the same place next time they open it. This feature gives users the choice of whether or not to save their current game state, including the items they hold, their current location, and their time within the time limit, and pick up where they left off the next time they want to play. 
+
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:16]: Awesome

File: docs/assignment3.md [hunk 6]
```diff
@@ -1 +1,205 @@
...
+| ID  | DP2  |
+|---|---|
+| **Design pattern**  | Command |
+| **Problem**  | We wanted a way to firstly separate our commands, since our Action class was getting complicated and hard to follow, and secondly be able to store the user's commands. |
+| **Solution**  | This design pattern allowed us to split each of the commands into separate classes, solving our first problem. It also allows us to store our commands like any other objects, solving the second problem. |
+| **Intended use**  | Now, when a user types a command, it calls the respective class's execute() method, which overrides the Command interface's execute() method. Now, each command can be stored as a separate object rather than just existing as a method in the Action class.  |
+| **Constraints**  | N/A |
+
+Maximum number of words for this section: 2000
+
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:39]: -Well motivated use of the command pattern to solve key issues with the project. Great!
-The decorator is used in a limited, niche way. What about additional gameplay alterations akin to removing the time limit? The problem statement is slightly off, you usually wouldn't use a design pattern to easily remove something, rather you would be interested in the added behavior that is provided from not having a time-limit i.e less stressful exploration.

File: docs/assignment3.md [hunk 7]
```diff
@@ -1 +1,205 @@
...
+The **ChatClient** class is what allows a player to communicate with other players through a chat room. The *serverAddress* holds the server IP that will be used. *TextField*, *messageArea*, and *frame* are attributes that are used to build the window that the chat room is in. *In* and *out* represent the text coming in and out of the chat room. 
+
+In our original class diagram, we did not have an Inventory class. This led to the methods such as getInventory and printInventory being scattered across multiple classes, making things harder to understand and use. Adding an extra class for the inventory made more sense, since there are a lot more necessary methods involving the inventory than we thought. We also added a ChatClient class, where we implemented our chat room feature for our game. Now, our game supports the command "chat," and it will open a chat room on your screen, which you can then talk to other players on. In addition, we added association names to the diagram to make the arrows easier to follow. We also added an abstract class to represent the JSON file that the game reads in. 
+
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:76]: Great changes overall, a couple of notes:
-JSON file shouldn't be abstract
-This stems from the decorator to some extend but the way you structured your player as an interface isn't very intuitive. You would want Player to provide some core functionality as an abstract class to all possible future modified iterations of Player(such as the time-decorated player)  like getInventory or getName, so you would only need to extend additional functionality instead of re-writing everything.

File: docs/assignment3.md [hunk 8]
```diff
@@ -1 +1,205 @@
...
+object - Amsterdam - is a collection and description of all of the rooms.
+
+In turn, the **Room** object connects to an **Item** object and an **Obstacle** object. The Item object, a coffee, is
+connected because it represents an item found in that particular room. The **Obstacle** object - in this case, a stop at
+the metro - represents a barrier to continued movement by the player that can be removed by using the correct item.
+
+Finally, the **Setting** object - Amsterdam - connects to a **TimeLimit** object and back to the **Room** of
+StationZuid. The room is one of many in the setting, while the **TimeLimit** will indicate how much time is left for the
+player to reach the goal.
+
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:103]: Good

File: docs/assignment3.md [hunk 9]
```diff
@@ -1 +1,205 @@
...
+If the player chooses to move, they must specify a location with a valid name which is adjacent to their currentLocation. If there is an obstacle located at the entry of this location, then they must possess the "key" in their inventory in order to use it and unlock the next room. If the player chooses to pick up an item, it must be one with a valid name that is in their currentLocation and is not already held. If the player puts down an item, similar conditions apply, but the item must already be in their inventory. 
+
+This cycle continues until either:
+The player runs out of time, i.e. their time is greater than the time limit specified in the JSON file and they are not yet located in the end room. In this case, a lose statement is printed out and the game is terminated.
+The player reaches their final destination, the “end room”, within the time limit. In this case, a win statement is printed out and the game finishes.
+
+
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:125]: The first fork node should have been a decision node, the rest If it was modeled in a composite state such as "exploring" with concurrent regions, it would've been good, otherwise, you can't use fork/join nodes.

File: docs/assignment3.md [hunk 10]
```diff
@@ -1 +1,205 @@
...
+If an item is picked up by the player, meaning the item now belongs to the player's inventory, the item changes state to "Item in player inventory". Upon entry to this state, the attribute *held* is set to true, meaning item is in the player's inventory and no longer in the room. Now, every time the player moves location, the item's *location* is updated as well.
+
+From the "Item in player inventory" state, the Item can be used, but only on the correct obstacle, which is represented by the arrow between the Item held and Item used states. 
+
+If the item is used from the "item in player inventory" state, the attribute *used* is set to true and the item is dropped from the inventory because it is no longer needed, thus terminating the item.
+
+Maximum number of words for this section: 4000
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:140]: Good

File: docs/assignment3.md [hunk 11]
```diff
@@ -1 +1,205 @@
...
+
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:156]: Great

File: docs/assignment3.md [hunk 12]
```diff
@@ -1 +1,205 @@
...
+This demo first shows the new Persistence feature by loading a JSON file that saved a previous game state, allowing the user to resume a previously played game. The demo then demonstrates the Obstacle feature by having to *use* an object before moving between rooms. The player in the demo loses the game by running out of time in order to demonstrate the Time Limit feature.
+
+One final note is that for a full game experience, the White Orchard JSON file inspired by the video game The Witcher 3
+should be played. The Amsterdam JSON mainly functions to illustrate the adaptability of the game to different settings
+as required by the assignment and is therefore not as extensive, but it is still fully playable.
+
+#### References
+
+The White_Orchard.json file is inspired by early levels of the video game The Witcher 3 by CD Projekt Red. Naturally,
+only the "plot" is referenced, all of the code is unique.
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:205]: genius

File: docs/assignment3.md [hunk 13]
```diff
@@ -1 +1,205 @@
...
+
+Another feature we added to the game was a live chatroom for players to communicate and help one another during gameplay. If the user enters the "chat" command, a **ChatClient** object is created that connects to a remote amazon ec2 server located at 13.58.146.33 on port 59001. The server code runs 7 days a week, from 8 to 16 UTC. If a user attempts to connect outside of chatroom hours they will be informed that the server is closed at that time.
+
+The main Java class needed for executing the system in the source code can be found at the directory location:
+software-design-vu/src/main/java/Main.java
+
+The Jar file can be found at the directory location:
+software-design-vu/out/artifacts/software_design_vu_2020_jar/software-design-vu-2020-jar.jar
+
```
Comment (NicholasKozanidis) [inline docs/assignment3.md:194]: Good work on the implementation, all features present and accounted for.

**Review by NicholasKozanidis — APPROVED** [review]
Good project overall, clearly attention was given to the feedback. One of the state machines is slightly off in it's use of fork/join nodes and out of the design patterns, the command was the biggest success while the use of decorator wasn't justified enough and felt like it solved an artificially created problem. Kudos for adding the chat.

File: src/main/java/Action.java (hunk 1 above)
Comment (PeterKementzey) [reply src/main/java/Action.java:6]: Indeed it seems like this is only a function, not a complete class, it could have been  a static function maybe, if you wanted to separate it from the rest of the code.

File: src/main/java/ChatClient.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+                this.frame,
+                "Choose a name for the chatroom:",
+                "Chatroom name selection",
+                JOptionPane.INFORMATION_MESSAGE
+        );
+    }
+
+    public void run() throws IOException {
+        try{
+            Socket socket = new Socket(serverAddress, 59001);
```
Comment (PeterKementzey) [inline src/main/java/ChatClient.java:46]: Avoid using 'magic numbers' I assume this is the port number. Make it a constant variable to improve readability and make the code explain itself.

File: src/main/java/ChatClient.java (hunk 2 above)
Comment (PeterKementzey) [reply src/main/java/ChatClient.java:54]: +1

File: src/main/java/Command.java [hunk 15]
```diff
@@ -0,0 +1,5 @@
+public interface Command {
```
Comment (PeterKementzey) [inline src/main/java/Command.java:1]: I think this is a nice example of using interfaces! :)

File: src/main/java/Inventory.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        for (Item item : items){
+            playerInventory = item.getNameFromItem() + ", " + playerInventory;
+        }
+
+        int unnecessaryComma = playerInventory.lastIndexOf(",");
+
+        return playerInventory.substring(0, unnecessaryComma);
+    }
+
+    public static String printInventory(ArrayList<Item> playerItems) {
```
Comment (PeterKementzey) [inline src/main/java/Inventory.java:28]: Why not use getStringInventory here instead of repeating the code? Actually I'm not  even sure this is a necessary function. Whenever the inventory has to be printed, the user of this class could just use the getStringInventory method and print the result wherever they want to.

File: src/main/java/Inventory.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        for (Item item : playerItems){
+            playerInventory = item.getNameFromItem() + ", " + playerInventory;
+        }
+        int unnecessaryComma = playerInventory.lastIndexOf(",");
+        playerInventoryFinal = "Your current inventory: " + playerInventory.substring(0, unnecessaryComma);
+
+        return playerInventoryFinal;
+    }
+
+    public boolean inventoryIsEmpty() {
```
Comment (PeterKementzey) [inline src/main/java/Inventory.java:41]: Since this is a method of the inventory class isEmpty as a name would be sufficient and obvious enough.

File: src/main/java/Inventory.java [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        if (items.isEmpty()) return true;
+        else return false;
+    }
+
+    public boolean isInInventory(String itemName) {
```
Comment (PeterKementzey) [inline src/main/java/Inventory.java:46]: contains would be a more concise name

File: src/main/java/Item.java [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+    public Item(String name, String location, String usage) {
+        this.name = name;
+        this.location = location;
+        this.usage = usage;
+        this.held = false;
+        this.used = false;
+    }
+
+    public String getNameFromItem() { return name; }
```
Comment (PeterKementzey) [inline src/main/java/Item.java:16]: minor comment: getNameOfItem or getName since it's part of the class anyway.

File: src/main/java/Main.java [hunk 20]
```diff
@@ -1,5 +1,259 @@
...
+        out.println(ANSI_RED + "\nWould you like to save your place, and return later? Answer: yes / no. \n" + ANSI_RESET);
+        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
+        String saveResponse;
+
+        try {
+            saveResponse = in.readLine();
+            if (saveResponse.equals("no")) {
+                out.println(ANSI_RED + "\nYou lose!\n" + ANSI_RESET);
+                exit(0);
+            } else if (saveResponse.equals("yes")) {
```
Comment (PeterKementzey) [inline src/main/java/Main.java:37]: This function is doing more than one thing, it could be split up into multiple parts, it is not clear from its name that it will save the game as well.

File: src/main/java/Main.java [hunk 21]
```diff
@@ -1,5 +1,259 @@
...
+                        + endRoom.getString("name") + ".\n You are currently located at " + startRoom.getString("name")
+                        + "." + Inventory.printInventory(startInventory) + ".\n Type "
+                        + ANSI_MAGENTA + "'help'" +ANSI_BLUE + " if you ever need help." + ANSI_RESET);
+            }
+        } catch(Exception e){
+            e.printStackTrace();
+        }
+    }
+
+    private static void whileGame(BufferedReader in, int timeLimitInt, JSONObject tomJsonObject) {
```
Comment (PeterKementzey) [inline src/main/java/Main.java:97]: This function name is not clear

File: src/main/java/Objective.java [hunk 22]
```diff
@@ -0,0 +1,23 @@
+import java.util.*;
+
+public class Objective {
```
Comment (PeterKementzey) [inline src/main/java/Objective.java:3]: It seems to me more of the functionality of this class should be implemented inside the class. It only has two accessor methods, the goal and result members are never used.

File: src/main/java/Obstacle.java [hunk 23]
```diff
@@ -0,0 +1,51 @@
+public class Obstacle {
```
Comment (PeterKementzey) [inline src/main/java/Obstacle.java:1]: Not much of it's functionality is implemented inside this class. clearObstacle cold for example check if the correct item is being used with the itemNeeded member instead of doing this outside the class.

File: src/main/java/Setting.java [hunk 24]
```diff
@@ -0,0 +1,9 @@
+import java.util.*;
+
+public class Setting {
```
Comment (PeterKementzey) [inline src/main/java/Setting.java:3]: I don't see the point of this class.

**Review by PeterKementzey — COMMENTED** [review]
Overall it seems you have worked a lot on the project. Good job! I could mostly comment on smaller implementation details and naming issues. Make sure to pay much attention to correct descriptive naming since that is a crucial quality of good code.

File: src/main/java/Command.java (hunk 15 above)
Comment (dominicsmorra) [reply src/main/java/Command.java:1]: ^ I have never used a chat client before and this was interesting to look at

File: src/main/java/Inventory.java [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+        String playerInventory = "";
+        String playerInventoryFinal;
+
+        for (Item item : playerItems){
+            playerInventory = item.getNameFromItem() + ", " + playerInventory;
+        }
+        int unnecessaryComma = playerInventory.lastIndexOf(",");
+        playerInventoryFinal = "Your current inventory: " + playerInventory.substring(0, unnecessaryComma);
+
+        return playerInventoryFinal;
```
Comment (dominicsmorra) [inline src/main/java/Inventory.java:38]: Seems a little redundant. There may be a better way to do this while only using one method instead of two.

File: src/main/java/Main.java [hunk 26]
```diff
@@ -1,5 +1,259 @@
...
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            } else {
+                out.println(ANSI_RED + "\nYou need to answer 'yes' or 'no'. Try again.\n" + ANSI_RESET);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
```
Comment (dominicsmorra) [inline src/main/java/Main.java:72]: This is a long method try splitting things up a little more for organizational purposes.

File: docs/assignment3.md (hunk 7 above)
Comment (dominicsmorra) [reply docs/assignment3.md:76]: Very descriptive overall but I agree there is no need for the json file to be abstract.

File: docs/assignment3.md (hunk 9 above)
Comment (dominicsmorra) [reply docs/assignment3.md:125]: ^

File: docs/assignment3.md (hunk 10 above)
Comment (dominicsmorra) [inline docs/assignment3.md:140]: Great Job!

File: docs/assignment3.md [hunk 27]
```diff
@@ -1 +1,205 @@
...
```
Comment (dominicsmorra) [inline docs/assignment3.md:150]: Very organized and well put together

File: docs/assignment3.md [hunk 28]
```diff
@@ -1 +1,205 @@
...
+The Jar file can be found at the directory location:
+software-design-vu/out/artifacts/software_design_vu_2020_jar/software-design-vu-2020-jar.jar
+
+![30-Second Demo](https://github.com/sarahpazik/software-design-vu/blob/Assignment3/Assignment3Demo1.mp4)
+This demo first shows the new Persistence feature by loading a JSON file that saved a previous game state, allowing the user to resume a previously played game. The demo then demonstrates the Obstacle feature by having to *use* an object before moving between rooms. The player in the demo loses the game by running out of time in order to demonstrate the Time Limit feature.
+
+One final note is that for a full game experience, the White Orchard JSON file inspired by the video game The Witcher 3
+should be played. The Amsterdam JSON mainly functions to illustrate the adaptability of the game to different settings
+as required by the assignment and is therefore not as extensive, but it is still fully playable.
+
```
Comment (dominicsmorra) [inline docs/assignment3.md:201]: Again very organized

File: docs/assignment1.md [hunk 29]
```diff
@@ -1,22 +1,18 @@
...
-The CommandParser class will handle the user’s commands, and determine whether an input is valid or not, given the object they are trying to use a command on. It will deal with any error handling that may be necessary.
-The Object class will contain all of the relevant information for different objects throughout the game, such as bikes, coffee, doors, etc. Each object will have a set of commands (from the Action class) that the user will be allowed to use on it.
+The Action class will allow the player to input valid commands, and then can see the consequences in the terminal. An example of this would be entering a new room (the action) using the keyword “enter”, and then seeing their new location and what items are now available to them.
+The Setting class will include generic information about what items, obstacles, etc. are in the game, as well as locations and directions from place to place. The location subclasses will hold the information for each specific room or location.
```
Comment (bjornkeyser) [inline docs/assignment1.md:10]: From the implementation it seems the Setting class only contains a variable rooms and does not need to be its own class.

File: docs/assignment3.md [hunk 30]
```diff
@@ -1 +1,205 @@
 # Assignment 3
+
+### Summary of changes of Assignment 2
+Author(s): Sarah, Sam
+
+* Made each of our actions into separate methods in the **Action** class to simplify out our previously long and complex switch statement
```
Comment (bjornkeyser) [inline docs/assignment3.md:6]: You took our input well and definitely improved readability

File: docs/assignment3.md [hunk 31]
```diff
@@ -1 +1,205 @@
...
+| **Intended use**  | We have a "broken clock" item in each game, and in our "pick" class/command, it checks whether the item picked up was the broken clock. If it was, the player's reference is changed to a TimeDecoratedPlayer rather than a RegularPlayer. This makes it so that the player's checkTime() method no longer counts down the time. |
+| **Constraints**  | N/A |
+
+| ID  | DP2  |
+|---|---|
+| **Design pattern**  | Command |
+| **Problem**  | We wanted a way to firstly separate our commands, since our Action class was getting complicated and hard to follow, and secondly be able to store the user's commands. |
+| **Solution**  | This design pattern allowed us to split each of the commands into separate classes, solving our first problem. It also allows us to store our commands like any other objects, solving the second problem. |
+| **Intended use**  | Now, when a user types a command, it calls the respective class's execute() method, which overrides the Command interface's execute() method. Now, each command can be stored as a separate object rather than just existing as a method in the Action class.  |
```
Comment (bjornkeyser) [inline docs/assignment3.md:35]: Good use of command design patterns

File: docs/assignment3.md [hunk 32]
```diff
@@ -1 +1,205 @@
...
+This cycle continues until either:
+The player runs out of time, i.e. their time is greater than the time limit specified in the JSON file and they are not yet located in the end room. In this case, a lose statement is printed out and the game is terminated.
+The player reaches their final destination, the “end room”, within the time limit. In this case, a win statement is printed out and the game finishes.
+
+
+#### Item State Machine
+
+The item state machine represents the internal behavior of the **Item** class.
+
+![Item State Machine Diagram](https://github.com/sarahpazik/software-design-vu/blob/Assignment3/Item%20State%20Machine%20Diagram%20(1).jpg)
```
Comment (bjornkeyser) [inline docs/assignment3.md:130]: Very well made and clear state machine diagram.

**Review by bjornkeyser — COMMENTED** [review]
Can't find much to criticize, well done and fun project.
